## Aula 03 - Filtragem Baseada em Conteúdo - Exercícios

In [1]:
import pandas as pd
import numpy as np

### Importar base de dados

In [72]:
import wget
!python3 -m wget https://github.com/mmanzato/MBABigData/raw/master/ml-20m-compact.tar.gz
!tar -xvzf ml-20m-compact.tar.gz

100% [....................................................] 65019041 / 65019041
Saved under ml-20m-compact.tar.gz
x dataset/
x dataset/tags_sample.csv
x dataset/.DS_Store
x dataset/movies_sample.csv
x dataset/genome-tags.csv
x dataset/ml-youtube.csv
x dataset/genome-scores.csv
x dataset/ratings_sample.csv


In [2]:
movies = pd.read_csv('./dataset/movies_sample.csv')
ratings = pd.read_csv('./dataset/ratings_sample.csv')
df = ratings[['userId', 'movieId', 'rating']]
df = df.merge(movies[['movieId', 'title']])
df

userId  movieId  rating                                     title
0           11     7481     5.0                         Enemy Mine (1985)
1          359     7481     4.0                         Enemy Mine (1985)
2          422     7481     3.0                         Enemy Mine (1985)
3          424     7481     3.0                         Enemy Mine (1985)
4          744     7481     3.0                         Enemy Mine (1985)
...        ...      ...     ...                                       ...
190616  131651   118606     3.5                               Dots (1940)
190617  134567   118606     3.5                               Dots (1940)
190618  131651    96257     1.0  Heroic Purgatory (Rengoku eroica) (1970)
190619  131651    98337     4.0                    97 Percent True (2008)
190620  131904   101666     2.0                 Night in Heaven, A (1983)

[190621 rows x 4 columns]

In [3]:
movies_tags = pd.read_csv('./dataset/tags_sample.csv')
movies_tags.head()

userId  movieId              tag  timestamp_y
0     279      916     Gregory Peck   1329962459
1     279      916      need to own   1329962471
2     279      916  romantic comedy   1329962476
3     279      916             Rome   1329962490
4     279      916          royalty   1329962474

In [4]:
map_users = {user: idx for idx, user in enumerate(df.userId.unique())}
map_items = {item: idx for idx, item in enumerate(df.movieId.unique())}
df['userId'] = df['userId'].map(map_users)
df['movieId'] = df['movieId'].map(map_items)
movies_tags['userId'] = movies_tags['userId'].map(map_users)
movies_tags['movieId'] = movies_tags['movieId'].map(map_items)
map_title = {}

for _, row in df.iterrows():
    map_title[row.movieId] = row.title


In [5]:
movies_tags.head()

userId  movieId              tag  timestamp_y
0   11080       34     Gregory Peck   1329962459
1   11080       34      need to own   1329962471
2   11080       34  romantic comedy   1329962476
3   11080       34             Rome   1329962490
4   11080       34          royalty   1329962474

### Divisão da base em treino e teste

In [6]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size=.2, random_state=2)

***Exercício 01:*** Aplique a filtragem baseada em conteúdo (ItemAttributeKNN do CaseRecommender) com as tags associadas aos filmes da base. Utilize Jaccard como métrica de similaridade e k=5 vizinhos para predição.

Documentação do ItemAttributeKNN: https://github.com/caserec/CaseRecommender/blob/master/caserec/recommenders/rating_prediction/item_attribute_knn.py

In [ ]:
# TODO

### Preparação para o exercício 2 - Download e extração de metadados multimídia

In [7]:
!python3 -m wget https://github.com/mmanzato/MBABigData/raw/master/ml-20m-features.tar.gz
! tar -xvzf ml-20m-features.tar.gz

100% [......................................................] 5996435 / 5996435
Saved under ml-20m-features.tar.gz
x features/
x features/m4infus_max_histogram_300_sn.arq
x features/mm_avg_histogram_100_sn.arq
x features/visual_histogram_100_sn.arq
x features/visual_histogram_50_sn.arq
x features/aural_histogram_50.arq
x features/mm_max_histogram_300.arq
x features/m4infus_max_histogram_50.arq
x features/mm_max_histogram_100.arq
x features/mm_max_histogram_50_sn.arq
x features/visual_histogram_100.arq
x features/visual_histogram_300.arq
x features/aural_histogram_100_sn.arq
x features/mm_avg_histogram_100.arq
x features/mm_max_histogram_100_sn.arq
x features/mm_sum_histogram_100_sn.arq
x features/mm_avg_histogram_300.arq
x features/visual_histogram_300_sn.arq
x features/mm_avg_histogram_300_sn.arq
x features/m4infus_max_histogram_100.arq
x features/m4infus_max_histogram_100_sn.arq
x features/mm_sum_histogram_50_sn.arq
x features/m4infus_max_histogram_300.arq
x features/mm_avg_histogram

In [7]:
import pickle

with open('./features/visual_histogram_50.arq', 'rb') as arq_visualHistograms:
    visualHistograms = pickle.load(arq_visualHistograms)
print('No. movies: ' + str(len(visualHistograms)))
print('Size of each word: ' + str(len(visualHistograms[0])))

No. movies: 433
Size of each word: 50


In [8]:
visualHistograms[0]

array([0.00948992, 0.00237248, 0.00355872, 0.01304864, 0.05338078,
       0.00711744, 0.05456702, 0.00355872, 0.0059312 , 0.00355872,
       0.0059312 , 0.00474496, 0.00711744, 0.04507711, 0.07591934,
       0.00355872, 0.1316726 , 0.05338078, 0.00830368, 0.        ,
       0.01304864, 0.00474496, 0.02253855, 0.00355872, 0.02728351,
       0.00237248, 0.02016607, 0.00830368, 0.0059312 , 0.00237248,
       0.00355872, 0.04033215, 0.01067616, 0.00237248, 0.00830368,
       0.05931198, 0.0059312 , 0.01067616, 0.02965599, 0.01423488,
       0.00711744, 0.02609727, 0.        , 0.00237248, 0.01423488,
       0.03084223, 0.02016607, 0.00118624, 0.08778173, 0.02253855])

***Exercício 02:*** Como visto, o algoritmo ItemAttributeKNN pode ser usado com diferentes tipos de metadados, como gêneros, tags e palavras no geral. Mais do que isso, podemos adaptá-lo também para que a similaridade entre itens seja feita com base em informações multimídia, como imagens, áudio, etc. 

A base de dados utilizada até o momento, ml-20m-compact.tar.gz possui, além das interações de usuários com filmes, uma série de arquivos que contém informações multimídia que foram extraídas dos trailers de cada filme. Esses arquivos estão condensados no zip ml-20m-features.tar.gz, o qual foi feito o download e extraído acima. 

Considere por exemplo o arquivo visual_histogram_50.arq. Ele possui 433 vetores (no. de filmes) de tamanho 50. Podemos pensar que cada vetor desse representa informações visuais (cor, brilho, imagem, etc.) que foram extraídas dos trailers de cada filme. 

Sua tarefa é usar esses vetores de características visuais no cálculo de similaridade entre os filmes, e em seguida, aplicar essas similaridades no algoritmo ItemAttributeKNN para gerar recomendações. 

Dica 1: para calcular a similaridade entre dois vetores pode-se usar o ângulo de cosseno (vide https://en.wikipedia.org/wiki/Cosine_similarity). 

Dica 2: é possível passar para o algoritmo ItemAttributeKNN a matriz de similaridade entre itens, por meio do parâmetro similarity_file=arquivo. Veja em: https://github.com/caserec/CaseRecommender/blob/master/caserec/recommenders/rating_prediction/item_attribute_knn.py

In [ ]:
# TODO

***Exercício 03:*** Implementar uma função que retorna a probabilidade de um item ser relevante para um usuário, considerando os gêneros dos filmes. Utilize métodos probabilísticos. 
- A partir do conjunto de treinamento, obter todas as interações do usuário u.
- Rotular as notas desse usuário como: item relevante se nota >=3 e não relevante se nota < 3. 
- Dado um item do conjunto de teste, aplicar o Teorema de Bayes com suavização de Laplace. Utilizar os gêneros associados. 
- Retornar se o item é ou não relevante, e em seguida, comparar o resultado com a nota real que esse usuário deu para o item (disponível no conjunto de teste).

In [ ]:
# TODO

***Exercício 04:*** No notebook de exemplos, existe uma implementação de Filtragem Baseada em Conteúdo usando Multi-Layer Perceptron como um regressor (MLPRegressor) que prevê a nota de cada usuário para filmes ainda não vistos. O algoritmo retorna uma lista de top K filmes com maiores notas. O treinamento é realizado utilizando as notas que o usuário deu para os filmes e seus respectivos gêneros.
- Usando a classe MLPClassifier (https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html), implemente uma versão que classifica os filmes não vistos como relevante ou não-relevante.
- No conjunto de dados, realize a binarização das notas, de modo que notas acima de 3 são relevantes e notas abaixo de 3 são não-relevantes.
- Retorne os top k filmes mais relevantes.
- Para um usuário qualquer da base (ou algum outro usuário fictício), analise subjetivamente a qualidade das recomendações, comparando os modelos MLPRegressor e MLPClassifier. 

In [ ]:
# TODO